src/data/list/to_finsupp.lean

```diff
@@ -0,0 +1,141 @@
+  [decidable_pred (λ (i : ℕ), nthd l i 0 ≠ 0)] (n : ℕ)
+
+/-- Indexing into a `l : list M`, as a finitely-supported function,
+where the support are all the indices within the length of the list
+that index to a non-zero value. Indices beyond the end of the list are sent to 0. -/
+def to_finsupp : ℕ →₀ M :=
+{ to_fun := λ i, nthd l i 0,
+  support := (finset.range l.length).filter (λ i, nthd l i 0 ≠ 0),
+  mem_support_to_fun := λ n, begin
+    simp only [ne.def, finset.mem_filter, finset.mem_range, and_iff_right_iff_imp],
+    contrapose!,
+    exact nthd_eq_default _ _
+  end }
+
+@[norm_cast] lemma coe_to_finsupp : (l.to_finsupp : ℕ → M) = λ i, l.nthd i 0 := rfl
+@[simp, norm_cast] lemma coe_to_finsupp_apply (i : ℕ) :
+  (l.to_finsupp : ℕ → M) i = l.nthd i 0 := rfl
+
+lemma to_finsupp_support :
+  l.to_finsupp.support = (finset.range l.length).filter (λ i, nthd l i 0 ≠ 0) :=
+rfl
+
+lemma to_finsupp_apply_lt (hn : n < l.length) :
+  l.to_finsupp n = l.nth_le n hn :=
+nthd_eq_nth_le _ _ _
+
+lemma to_finsupp_apply_le (hn : l.length ≤ n) :
+  l.to_finsupp n = 0 :=
+nthd_eq_default _ _ hn
+
+@[simp] lemma to_finsupp_nil [decidable_pred (λ (i : ℕ), nthd ([] : list M) i 0 ≠ 0)] :
+  to_finsupp ([] : list M) = 0 :=
+by { ext, simp }
+
+lemma to_finsupp_singleton (x : M)
+  [decidable_pred (λ (i : ℕ), nthd [x] i 0 ≠ 0)] :
+  to_finsupp [x] = finsupp.single 0 x :=
+begin
+  ext ⟨_|i⟩;
+  simp [finsupp.single_apply, (nat.zero_lt_succ _).ne]
+end
+
+@[simp] lemma to_finsupp_cons_apply_zero (x : M) (xs : list M)
+  [decidable_pred (λ (i : ℕ), nthd (x :: xs) i 0 ≠ 0)] :
+  (x :: xs).to_finsupp 0 = x := rfl
+
+@[simp] lemma to_finsupp_cons_apply_succ (x : M) (xs : list M) (n : ℕ)
+  [decidable_pred (λ (i : ℕ), nthd (x :: xs) i 0 ≠ 0)]
+  [decidable_pred (λ (i : ℕ), nthd xs i 0 ≠ 0)] :
+  (x :: xs).to_finsupp n.succ = xs.to_finsupp n := rfl
+
+lemma to_finsupp_cons_eq_single_add_emb_domain
+  {R : Type*} [add_zero_class R] (x : R) (xs : list R)
+  [decidable_pred (λ (i : ℕ), nthd (x :: xs) i 0 ≠ 0)]
+  [decidable_pred (λ (i : ℕ), nthd xs i 0 ≠ 0)] :
+  to_finsupp (x :: xs) = finsupp.single 0 x +
+    (to_finsupp xs).emb_domain ⟨nat.succ, nat.succ_injective⟩ :=
+begin
+  ext (_|i),
+  { simp only [nat.nat_zero_eq_zero, to_finsupp_cons_apply_zero, finsupp.coe_add,
+               pi.add_apply, finsupp.single_eq_same],
+    rw finsupp.emb_domain_notin_range,
+    { exact (add_zero _).symm },
+    { simp } },
+  { simp only [to_finsupp_cons_apply_succ, finsupp.coe_add, pi.add_apply],
+    have hi : i.succ = (⟨nat.succ, nat.succ_injective⟩ : ℕ ↪ ℕ) i := rfl,
+    rw [finsupp.single_apply_eq_zero.mpr, zero_add, hi, finsupp.emb_domain_apply],
+    simp }
+end
+
+lemma to_finsupp_concat_eq_to_finsupp_add_single
+  {R : Type*} [add_zero_class R] (x : R) (xs : list R)
+  [decidable_pred (λ (i : ℕ), nthd (xs ++ [x]) i 0 ≠ 0)]
+  [decidable_pred (λ (i : ℕ), nthd xs i 0 ≠ 0)] :
+  to_finsupp (xs ++ [x]) = to_finsupp xs + finsupp.single xs.length x :=
```

Comment:
You could use `list.concat` here.

Reply:
The simp normal form of `list.concat` is this "append a singleton" style.